### DISCUSSION TITLE
Linker error in Post-02 with custom 32-bit bootloader

### DISCUSSION THREAD

**Discussion by sadkris** [phil-opp/blog_os#1300] [General]
Hi! I'm using i686 to compile for the OS I'm working with. My target specification is as follows:
```
{
  "llvm-target": "i686-unknown-none",
  "data-layout": "e-m:e-p:32:32-p270:32:32-p271:32:32-p272:64:64-i128:128-f64:32:64-f80:32-n8:16:32-S128",
  "arch": "x86",
  "target-endian": "little",
  "target-pointer-width": "32",
  "target-c-int-width": "32",
  "os": "none",
  "executables": true,
  "linker-flavor": "ld.lld",
  "linker": "rust-lld",
  "panic-strategy": "abort",
  "disable-redzone": true,
  "features": "+soft-float,-sse,-mmx"
}
```

I'm currently compiling the example code from Post-02 on printing bytes to VRAM. The script successfully compiles in Rust, but when linking it with my bootloader externally, it produces the following error:
```
`i686-elf-ld: Rust_Test.o.3: in function `_start':
/home/kali/Rust-Test/src/main.rs:20:(.text._start+0x20): undefined reference to `core::slice::<impl [T]>::iter'
i686-elf-ld: /home/kali/Rust-Test/src/main.rs:20:(.text._start+0x34): undefined reference to `core::iter::traits::iterator::Iterator::enumerate'
i686-elf-ld: /home/kali/Rust-Test/src/main.rs:20:(.text._start+0x4b): undefined reference to `<I as core::iter::traits::collect::IntoIterator>::into_iter'
i686-elf-ld: /home/kali/Rust-Test/src/main.rs:20:(.text._start+0x72): undefined reference to `<core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::next'
i686-elf-ld: /home/kali/Rust-Test/src/main.rs:22:(.text._start+0x123): undefined reference to `core::panicking::panic'
i686-elf-ld: /home/kali/Rust-Test/src/main.rs:23:(.text._start+0x150): undefined reference to `core::panicking::panic'
i686-elf-ld: /home/kali/Rust-Test/src/main.rs:23:(.text._start+0x18a): undefined reference to `core::panicking::panic'
```

Would really appreciate some help! My main.rs file is as follows:
```
#![no_std] // don't link the Rust standard library
#![no_main] // disable all Rust-level entry points

use core::panic::PanicInfo;

/// This function is called on panic.
#[panic_handler]
fn panic(_info: &PanicInfo) -> ! {
    loop {}
}

static HELLO: &[u8] = b"Hello World!";

#[no_mangle]
pub extern "C" fn _start() -> ! {
    let vga_buffer = 0xb8000 as *mut u8;

    for (i, &byte) in HELLO.iter().enumerate() {
        unsafe {
            *vga_buffer.offset(i as isize * 2) = byte;
            *vga_buffer.offset(i as isize * 2 + 1) = 0xb;
        }
    }

    loop {}
}
```

**Comment by bjorn3** [comment]
How are you linking the code? It looks like you are invoking the linker yourself rather than having rustc invoke it for you. You may be missing necessary linker flags.

**Comment by sadkris** [comment]
> How are you linking the code? It looks like you are invoking the linker yourself rather than having rustc invoke it for you. You may be missing necessary linker flags.

After compiling it in rustc, I'm using the command `i686-elf-ld -o tmp/kernel.bin -Ttext 0x1000 tmp/kernel_entry.o tmp/kernel.o --oformat binary` with the compiled .o file being kernel.o, and kernel_entry being an assembly script to import and run _start.

**Comment by bjorn3** [comment]
How did you get that `kernel.o` file? If you literally passed `--emit obj`, that will not work. That will not include the machine code for dependencies. What you could do however is compile with the staticlib crate type and then link the resulting `libkernel.a`. This bundles all dependencies.

**Comment by sadkris** [comment]
What would the config need to look like in order to do that?

**Comment by bjorn3** [comment]
In `Cargo.toml` you can add

```toml
[lib]
crate-type = ["staticlib"]
```

**Comment by sadkris** [comment]
Recompiling it with that change, as well as changing the file name from main to lib, then copying the .a file from `target/i686/debug`, it produced a linking error since it was using a GOT with PIE enabled. When I disabled PIE, the script compiled correctly and linked properly, but running it in qemu gets stuck in a boot loop of flashing the "SeaBIOS" header as well as "Booting from hard disk", then going to a blank screen, then repeating.

**Comment by phil-opp** [comment]
The bootloader crate only supports booting x86_64 kernels, not 32-bit kernels.

**Comment by sadkris** [comment]
> The bootloader crate only supports booting x86_64 kernels, not 32-bit kernels.

I am not using the bootloader crate at all. As I mentioned before, I'm using my own bootloader

**Comment by phil-opp** [comment]
Ah, I missed that! In that case, you probably have to debug your bootloader. You can either connect gdb and step through the assembly to see where it fails, or you use QEMU's `-d int` flag to see which exceptions occur where. If course you can also insert endless loops at various stages in your bootloader to narrow down which part causes the crash.

**Comment by phil-opp** [comment]
(I migrated this issue to a discussion because it is not an issue of the blog itself. I also updated the title to clarify that this is related to a custom 32-bit bootloader.)
